File: src/mock/data/building-map.ts
import * as RomiCore from '@osrf/romi-js-core-interfaces';
import airportPngUrl from './airport_terminal.png';
import officePngUrl from './office.png';
import officeMap from './building-map-office';
import airportMap from './building-map-airport';

export default async function buildingMap(): Promise<RomiCore.BuildingMap> {
  let officePng: Uint8Array;
  let airportPng: Uint8Array;
  if (process.env.NODE_ENV === 'test') {
    const fs = await import('fs');
    officePng = fs.readFileSync(`${__dirname}/office.png`);
    airportPng = fs.readFileSync(`${__dirname}/airport_terminal.png`);
  } else {
    const officePngResp = await fetch(officePngUrl);
    officePng = new Uint8Array(await officePngResp.arrayBuffer());
    const airportPngResp = await fetch(airportPngUrl);
    airportPng = new Uint8Array(await airportPngResp.arrayBuffer());
  }

  return {
    name: 'test building',
    levels: [
      {
        name: 'L1',
        elevation: 0,
        images: [
          {
            data: officePng,
            encoding: 'png',
            name: 'office',
            scale: 0.008465494960546494,
            x_offset: 0,
            y_offset: 0,
            yaw: 0,
          },
        ],
        doors: officeMap.doors,
        places: officeMap.places,
        nav_graphs: [],
        wall_graph: {
          name: 'wallgraph',
          vertices: [],
          edges: [],
          params: [],
        },
      },
      {
        name: 'Airport Terminal',
        elevation: 10,
        images: [
          {
            data: airportPng,
            encoding: 'png',
            name: 'airport_terminal',
            scale: 0.08212187141180038,
            x_offset: 0,
            y_offset: 0,
            yaw: 0,
          },
        ],
        doors: airportMap.doors,
        places: airportMap.places,
        nav_graphs: [],
        wall_graph: {
          name: 'wallgraph',
          vertices: [],
          edges: [],
          params: [],
        },
      },
    ],
    lifts: officeMap.lifts
  };
}

File: packages/api-server/api_server/models/building_map.py
from pydantic import BaseModel

from . import tortoise_models as ttm
from .ros_pydantic import rmf_building_map_msgs


class AffineImage(BaseModel):
    name: str
    x_offset: float
    y_offset: float
    yaw: float
    scale: float
    encoding: str
    data: str


class Level(BaseModel):
    name: str
    elevation: float
    images: list[AffineImage]
    places: list[rmf_building_map_msgs.msg.Place]
    doors: list[rmf_building_map_msgs.msg.Door]
    nav_graphs: list[rmf_building_map_msgs.msg.Graph]
    wall_graph: rmf_building_map_msgs.msg.Graph


class BuildingMap(BaseModel):
    name: str
    levels: list[Level]
    lifts: list[rmf_building_map_msgs.msg.Lift]

    @staticmethod
    def from_tortoise(tortoise: ttm.BuildingMap) -> "BuildingMap":
        return BuildingMap(**dict(tortoise.data))


class FireAlarmTriggerState(BaseModel):
    unix_millis_time: int
    trigger: bool
